CDAP-7930 Fix incorrect metadata search total

Denton-L opened the pull request:
Fixes the following:

* An incorrect total would be returned if an offset > number of results
  was specified while doing a query with non-default sort order.

* If offset > number of results, limit does not restrict the number of
  results appropriately.

* ArrayList was being used where LinkedList would be more efficient.

* A negative offset or limit was unchecked and could potentially cause
  problems.

JIRA: https://issues.cask.co/browse/CDAP-7930
Bamboo: http://builds.cask.co/browse/CDAP-DUT5420-1

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java:
```diff
@@ -442,9 +440,10 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
     return new MetadataSearchResponse(
-      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, total,
-      addMetadataToEntities(sortedEntities, systemMetadata, userMetadata), cursors, showHidden
+      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, resultRecords.size(),
```
Is this correct? If in total there are 100 records, if request offset=0 and limit=10, then `resultRecords.size()` is only `10` instead of `100`, right? I thought total means total number of records, i.e. something won't change regardless of what `limit` and `offset` is.

Denton-L replied:
Originally, it `total` would be 100. In this change, the number will now be `10`. I'm not sure if this is the "correct" behaviour but this is what the JIRA requested.

If this is incorrect, I can close the JIRA as won't fix.

chtyim replied:
@bdmogal can you clarify the requirement? The JIRA is not very clear to me.

bdmogal replied:
This seems wrong. Search should **always** return the total number of entities. @anwar6953 suspected that there is a bug in the total, when you specify *offset* > *number of results*. That's what needs to be verified, I think.

Like @chtyim said, when there are 100 records, if offset = 0 and limit = 10, total should remain 100, because that helps the UI calculate the number of pages.

Denton-L replied:
I messaged him on Slack before starting this ticket and he said this:

>if giving an offset of 10, that means it should skip the first 10 entries
>and also not include those in the ‘total’ number

So I guess this ticket should be closed as won't fix because we are, indeed, returning the correct total.

anwar6953 replied:
Yea, I suspect an issue to be there when the offset is a large number (larger than the number of actual results).

anwar6953 replied:
@Denton-L can you verify what the 'total' is, when specifying an offset that is much larger than the actual total?

Denton-L replied:
It seems to return the same total, regardless of what offset you specify.

Denton-L replied:
Actually, revisiting it, I think there actually is a bug for when you provide a `sort` parameter... I'll look into it.

anwar6953 replied:
That's with this fix, right?
And without the fix, the total is much higher (depending on the offset provided)?

Denton-L replied:
Currently, without the fix, the query will return a correct total for any offset given that the `sort` param is not given. If the `sort` param is given, the total is incorrect without the fix.

I am currently writing a new patch to fix this because I think I've found the root of the problem.

Denton-L committed:
CDAP-7930 Fix an incorrect total being reported

Fixes the following:

* An incorrect total would be returned if an offset > number of results
  was specified while doing a query with non-default sort order.

* If offset > number of results, limit does not restrict the number of
  results appropriately.

* ArrayList was being used where LinkedList would be more efficient.

* A negative offset or limit was unchecked and could potentially cause
  problems.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java:
```diff
@@ -468,7 +468,7 @@ private Metadata getSnapshotBeforeTime(NamespacedEntityId targetId, long timeMil
     }
 
-    List<ImmutablePair<byte [], byte []>> fuzzyKeys = new ArrayList<>();
+    List<ImmutablePair<byte [], byte []>> fuzzyKeys = new LinkedList<>();
```
Why the change? Also, since we know the expected size (`targetIds.size()`), we should use `new ArrayList<>(targetIds.size())`.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java:
```diff
@@ -563,15 +563,14 @@ private MetadataEntry convertRow(Row row) {
   private SearchResults searchByDefaultIndex(String namespaceId, String searchQuery,
                                              Set<EntityTypeSimpleName> types, boolean showHidden) {
-    List<MetadataEntry> results = new ArrayList<>();
+    List<MetadataEntry> results = new LinkedList<>();
```
Why the change?

Denton-L replied:
Since we're not doing any random-access and we don't know the size of the results prior to creation, I feel like it would be more efficient to use a `LinkedList` over an `ArrayList`. Please let me know if I'm wrong about this.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java:
```diff
@@ -398,44 +399,44 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
                          showHidden);
       results.addAll(searchResults.getResults());
       cursors.addAll(searchResults.getCursors());
+      allResults.addAll(searchResults.getAllResults());
```
Why storing all results??

Denton-L replied:
We need to get all of the results so that we can count them in the total, I believe.

Denton-L commented:
@chtyim I've tried to address all of the comments.

sreevatsanraman commented:
LGTM
